frontend/src/components/widgets/FileUploader/FileUploader.tsx

```diff
@@ -306,30 +351,36 @@ class FileUploader extends React.PureComponent<Props, State> {
   }
 
   /** Remove the file with the given ID from `state.files`. */
-  private removeFile = (idToRemove: string): void => {
+  private removeFile = (idToRemove: number): void => {
     this.setState(state => ({
       files: state.files.filter(file => file.id !== idToRemove),
     }))
   }
 
+  /**
+   * Return the file with the given ID, if one exists.
+   */
+  private getFile = (fileId: number): UploadFileInfo | undefined => {
+    return this.state.files.find(file => file.id === fileId)
+  }
+
   /** Replace the file with the given id in `state.files`. */
-  private updateFile = (
-    idToUpdate: string,
-    replacement: UploadFileInfo
-  ): void => {
-    this.setState(state => ({
-      files: state.files.map(file =>
-        file.id === idToUpdate ? replacement : file
-      ),
-    }))
+  private updateFile = (curFileId: number, newFile: UploadFileInfo): void => {
+    this.setState(curState => {
+      return {
+        files: curState.files.map(file =>
```

Comment:
Any state changes that modify existing state needs to use the function mechanism, because state updates can be coalesced by React. Meaning: it's possible to have multiple state updates applied before a single render. 

(So if you have e.g. a counter that you're incrementing, and you did something like `this.setState({ counter: this.state.counter + 1})` multiple times, you could have state updates overwriting each other and end up with the wrong value.)

Reply:
Gotcha!